File: library/src/main/java/com/zdong/stateviewmanager/StateManagerView.java
package com.zdong.stateviewmanager;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

import com.zdong.stateviewmanager.creator.StateLoader;
import com.zdong.stateviewmanager.creator.StateViewRepository;
import com.zdong.stateviewmanager.mananger.StateActionListener;
import com.zdong.stateviewmanager.mananger.StateManager;
import com.zdong.stateviewmanager.mananger.StateViewChanger;
import com.zdong.stateviewmanager.state.IStateView;
import com.zdong.stateviewmanager.state.StateProperty;

/**
 *
 * @author zoudong
 */
public class StateManagerView extends FrameLayout implements StateViewChanger, StateLoader<View> {

    StateManager mStateManager;

    public StateManagerView(Context context) {
        this(context, null);
    }

    public StateManagerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public StateManagerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public StateManagerView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    private void init() {
        mStateManager = StateManager.newInstance(getContext(), new StateViewRepository(getContext()), this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        if (getChildCount() > 1) {
            try {
                throw new IllegalStateException("StateManagerView can have only one direct child");

            } catch (IllegalStateException e) {
                e.printStackTrace();
            }
        } else if (getChildCount() == 1) {
            loadCoreView(getChildAt(0));
        }
    }
    private void loadCoreView(View child) {
        if (child != null) {
            mStateManager.setContentView(child);
        }
    }
    @Override
    public boolean addState(IStateView changger) {
        return mStateManager.addState(changger);
    }

    @Override
    public boolean removeState(String state) {
        return mStateManager.removeState(state);
    }

    @Override
    public View getStateView(String state) {
        return mStateManager.getStateView(state);
    }

    @Override
    public boolean showState(String state) {
        return mStateManager.showState(state);
    }

    @Override
    public boolean showState(StateProperty state) {
        return mStateManager.showState(state);
    }

    @Override
    public String getState() {
        return mStateManager.getState();
    }

    @Override
    public void setStateActionListener(StateActionListener listener) {
        mStateManager.setStateActionListener(listener);
    }

    @Override
    public void onDestoryView() {
        mStateManager.onDestoryView();
    }


    public static class Builder {
        StateManagerView mStateManagerView;

        public Builder(Context context) {
            mStateManagerView = new StateManagerView(context);
        }

        public Builder wrapper(View view) {
            mStateManagerView.mStateManager.setContentView(view);
            return this;
        }

        public StateViewChanger builder() {
            return mStateManagerView;
        }
    }

}

File: library/src/main/java/com/zdg/stateviewmanager/StateManagerView.kt
package com.zdg.stateviewmanager

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout

import com.zdg.stateviewmanager.creator.StateLoader
import com.zdg.stateviewmanager.creator.StateViewRepository
import com.zdg.stateviewmanager.mananger.StateActionListener
import com.zdg.stateviewmanager.mananger.StateManager
import com.zdg.stateviewmanager.mananger.StateViewChanger
import com.zdg.stateviewmanager.state.IStateView
import com.zdg.stateviewmanager.state.StateProperty

/**
 *
 * @author zoudong
 */
class StateManagerView : FrameLayout, StateViewChanger, StateLoader<View> {

    var mStateManager: StateManager = StateManager.newInstance(context, StateViewRepository(context), this)

    override val state: String
        get() = mStateManager.state

    @JvmOverloads
    constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)


    override fun onFinishInflate() {
        super.onFinishInflate()
        if (childCount > 1) {
            try {
                throw IllegalStateException("StateManagerView can have only one direct child")
            } catch (e: IllegalStateException) {
                e.printStackTrace()
            }

        } else if (childCount == 1) {
            loadCoreView(getChildAt(0))
        }
    }

    private fun loadCoreView(child: View?) {
        if (child != null) {
            mStateManager.setContentView(child)
        }
    }
    override fun addState(iStateView: IStateView<StateProperty>): Boolean {
        return mStateManager.addState(iStateView)
    }

    override fun removeState(state: String): Boolean {
       return mStateManager.removeState(state)
    }

    override fun getStateView(state: String): View? {
        return mStateManager.getStateView(state)
    }

    override fun showState(state: String): Boolean {
        return mStateManager.showState(state)
    }

    override fun showState(state: StateProperty): Boolean {
        return mStateManager.showState(state)
    }

    override fun setStateActionListener(listener: StateActionListener) {
        mStateManager.setStateActionListener(listener)
    }

    override fun onDestoryView() {
        mStateManager.onDestoryView()
    }


    class Builder(context: Context) {
        internal var mStateManagerView: StateManagerView = StateManagerView(context)

        fun wrapper(view: View): Builder {
            mStateManagerView.mStateManager.setContentView(view)
            return this
        }

        fun builder(): StateViewChanger {
            return mStateManagerView
        }
    }

}
